### ISSUE TITLE
_IncompatibleKeys returned from PeftModel.load_adapter

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/peft#1323]
Hello all, and thank you for the great work being done here!

I am experimenting with LoRA adapters. However, when trying to load a pretrained adapter, I recieive an _IncompatibleKeys object containing a very large amount of parameters in the `missing_keys` field. I note there are no `unexpected_keys` reported.

Here is some minimum reproducible code (which I run in a Jupyter notebook).
```python
from peft.mixed_model import PeftMixedModel

model = AutoModelForCausalLM.from_pretrained(...)

model = PeftMixedModel.from_pretrained(model,
                                       './Mistral_v120Zephyr/checkpoint-30780/',
                                       'Bioinspired', False)

model.load_adapter('./Mistral_v121Zephyr/checkpoint-37378/', 'CoT')
```

And here is the output, which I have truncated:
```python
_IncompatibleKeys(missing_keys=['base_model.model.model.embed_tokens.weight', 
'base_model.model.model.layers.0.self_attn.q_proj.base_layer.weight', 
'base_model.model.model.layers.0.self_attn.q_proj.lora_A.CoT.weight', ...])
```

After doing some digging, I see that _IncompatibleKeys is returned from `torch.nn.Module.load_state_dict` when (in one case) the state dicts to not match. Is this normal behavior for `load_adapter`, or am I doing something wrong? I would appreciate any help!

Edit:
Please find the full file here: 
[incompatible_keys.txt](https://github.com/huggingface/peft/files/13842137/incompatible_keys.txt)

I noticed that some parameters are `base_model.model.model` (as above), and some are `base_model.model`. Could this be a problem?

**Comment by EricLBuehler** [comment]
Yes, I plan on mixing LoRA adapters, so `PeftMixedModel` is mandatory from my understanding. Am I using the API incorrectly to cause this "error"?

**Comment by BenjaminBossan** [comment]
If you want to mix LoRA adapters, `PeftModel` is sufficient. `PeftMixedModel` is for mixing _different types of adapters_ (e.g. LoRA + AdaLoRA + LoKr).

**Comment by BenjaminBossan** [comment]
> The issue persists with that approach, too.

Do you get the incompatible keys from the first load or from the 2nd? Can you check if it still works, despite the incompatible keys?

**Comment by EricLBuehler** [comment]
> If you want to mix LoRA adapters, PeftModel is sufficient
Thanks for the clarification, however the issue persists when using `PeftModel`.

> Do you get the incompatible keys from the first load or from the 2nd? 
It is the 2nd load. 

> Can you check if it still works, despite the incompatible keys?
How can I check if it still works?

**Comment by BenjaminBossan** [comment]
To check if both `state_dict`s were loaded, you could print the model. There should be LoRA layers (probably `lora.Linear`) that have `ModuleDict`s with keys for both of your adapters, i.e. `'Bioinspired'` and `'CoT'`. But this just indicates that something was loaded, not that it works correctly. To verify the latter, you first need to activate the 2nd adapter:

```python
model.set_adapter("CoT")
```

Next, ideally, you could run a couple of predictions and check if they conform to expectations. Perhaps, you have a dataset used to train the 2nd adapter, which can be used to check the scores of the loaded model?

**Comment by EricLBuehler** [comment]
After printing the model, it appears that the `state_dict`s were loaded. Additionally, a subsequent SVD merging operation worked too. I have been developing on the same machine with a constant PyTorch version.

Is it safe to ignore the `_IncompatibleKeys` warning if the `state_dict`s are loaded and the SVD merging works? Alternatively, would you recommend that I open an issue with PyTorch for this?

**Comment by BenjaminBossan** [comment]
> Is it safe to ignore the `_IncompatibleKeys` warning if the `state_dict`s are loaded and the SVD merging works?

I have another proposal to test this quickly. Presumably, it's always the 2nd adapter that reports the incompatible keys. So if you load `CoT` first, it should load fine. Could you try that and then check if the predictions from that adapter are the same as you would get when you load `CoT` second (as in your first example) and then activate it with `set_adapter` (ensuring eval mode and same random seed)? If the two lead to the same results, we can be pretty sure that loading worked correctly.

> Alternatively, would you recommend that I open an issue with PyTorch for this?

Let's not jump ahead yet :) If we can establish that this is indeed a false positive, we first should work out where it comes from before creating an issue on PyTorch.

**Comment by SumanthRH** [comment]
I can reproduce the same `load_result` with `PeftModel`. I believe this is intended behaviour and not an issue. To quickly see the same:

```
from peft import get_peft_model, PeftModel, LoraConfig
from transformers import AutoModelForCausalLM
model =  AutoModelForCausalLM.from_pretrained("hf-internal-testing/tiny-random-gpt2")
peft_model = get_peft_model(model, LoraConfig())
load_result = peft_model.load_adapter("SumanthRH/test_config", adapter_name="adapter2") # a test adapter I've created
print(load_result)
```

The `_IncompatibleKeys` comes up because of [this line](https://github.com/huggingface/peft/blob/6451cbd70c187dc7baeadeed701851ac7bdd6ec2/src/peft/utils/save_and_load.py#L211) in `set_peft_model_state_dict` :
```
load_result = model.load_state_dict(peft_model_state_dict, strict=False)
```

Here, `peft_model_state_dict` is the state dict containing our new adapter weights (and ONLY the new adapter weights). Internally, PEFT will rename the state dict entries to use the new adapter name `adapter2`. So in my case, it looks like:
```
'base_model.model.transformer.h.0.attn.c_attn.lora_A.adapter2.weight', 'base_model.model.transformer.h.0.attn.c_attn.lora_B.adapter2.weight', .......
```

When you call `load_adapter`, a new adapter is initialized first using `add_adapter` [here](https://github.com/huggingface/peft/blob/6451cbd70c187dc7baeadeed701851ac7bdd6ec2/src/peft/peft_model.py#L687). Next, when we load the state dict for the new adapter using `model.load_state_dict`, PyTorch finds that all the entries for the new adapter are present, but the entries for the base model and the old adapter are not present. Hence you get a `_IncompatibleKeys` load result with a long list of missing keys (and no unexpected keys).

**Comment by BenjaminBossan** [comment]
Thanks a lot @SumanthRH for digging into this. Your explanation makes a lot of sense. IMO the name `_IncompatibleKeys` is a bit imprecise in this context, it would be better to call them missing keys, but we can't really do anything about that.

I'll close the issue then, feel free to re-open if anything new comes up @EricLBuehler.

**Comment by EricLBuehler** [comment]
@BenjaminBossan, @SumanthRH thank you for the help! That answer makes sense.

**Comment by EricLBuehler** [comment]
Is this a Pytorch issue, then?
> The `_IncompatibleKeys` comes up because of [this line](https://github.com/huggingface/peft/blob/6451cbd70c187dc7baeadeed701851ac7bdd6ec2/src/peft/utils/save_and_load.py#L211) in `set_peft_model_state_dict` :
> 
> ```
> load_result = model.load_state_dict(peft_model_state_dict, strict=False)
> ```
> 
> Here, `peft_model_state_dict` is the state dict containing our new adapter weights (and ONLY the new adapter weights). Internally, PEFT will rename the state dict entries to use the new adapter name `adapter2`. So in my case, it looks like:
> 
> ```
> 'base_model.model.transformer.h.0.attn.c_attn.lora_A.adapter2.weight', 'base_model.model.transformer.h.0.attn.c_attn.lora_B.adapter2.weight', .......
> ```
> 
> When you call `load_adapter`, a new adapter is initialized first using `add_adapter` [here](https://github.com/huggingface/peft/blob/6451cbd70c187dc7baeadeed701851ac7bdd6ec2/src/peft/peft_model.py#L687). Next, when we load the state dict for the new adapter using `model.load_state_dict`, PyTorch finds that all the entries for the new adapter are present, but the entries for the base model and the old adapter are not present. Hence you get a `_IncompatibleKeys` load result with a long list of missing keys (and no unexpected keys).

**Comment by SumanthRH** [comment]
@EricLBuehler it's not a Pytorch issue! Pytorch is behaving as expected when we called `load_state_dict`. An `_InCompatibleKeys` object is returned because the model state dict and `peft_model_state_dict` have a different number of keys. It is up to the user to make sense of the result, and raise an error if needed. In this case, it is expected behavior because `peft_model_state_dict` is supposed to have only the key, value pairs for the new adapter.

**Comment by SumanthRH** [comment]
> Additionally, the missing_keys attribute contains entries from both adapters, so does that imply that keys from both are not found in the new state dict, when only one should be present
Oh this was different for me. But let me check.

**Comment by EricLBuehler** [comment]
Ok, thanks!

**Comment by Jwmc999** [comment]
To anyone who have this issue, I hope my case helps.
For me, it was whether the peftmodel I am downloading went both training and evaluation or not.
```python
# SFT training args
training_arguments = TrainingArguments(
        output_dir="./results",
        num_train_epochs=1,
        per_device_train_batch_size=4,
        gradient_accumulation_steps=1,
        gradient_checkpointing=True, 
        evaluation_strategy="steps",
        eval_steps=1000,
        logging_steps=25,
        optim="paged_adamw_8bit",
        learning_rate=2e-4,
        lr_scheduler_type="linear",
        warmup_steps=10,
        report_to="tensorboard",
        max_steps=-1, 
)

# DPO training args
training_args = TrainingArguments(
    per_device_train_batch_size=2,
    max_steps=125, #running 125 steps (1 epoch) for demo purpose
    gradient_accumulation_steps=4,
    gradient_checkpointing=True,
    learning_rate=5e-5,
    lr_scheduler_type="cosine",
    save_strategy="no",
    logging_steps=1,
    output_dir=new_model,
    optim="paged_adamw_32bit",
    warmup_steps=5,
    remove_unused_columns=False,
)
```
I removed `evaluation_strategy` and `eval_steps` from SFT training args then `_IncompatibleKeys returned from PeftModel.load_adapter` disappeared
